### PR TITLE
Encoding the email address as asn.1 ia5

### PR DESCRIPTION
**Pull request by hwine**
lines with '//hwine' are additions for use when running locally (i.e.
without nCipher present). Code needing to be commented out is not
marked.

### CODE REVIEW

**Comment by hwine** [comment]
Sample CSR produced by the above code:
```
+&{[] [] [] [] 0 [] SHA256-RSA 0 <nil> 1.2.840.113549.1.9.1=#04243022162072656c656173652b636572746966696361746573406d6f7a696c6c612e636f6d,CN=Mozilla Corporation,OU=Release Engineering,O=Mozilla Corporation,L=Mountain View,ST=California,C=US [] [] [] [Mozilla Corporation] [] [] []}
-----BEGIN CERTIFICATE REQUEST-----
MIIDQDCCAigCAQAwgckxCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpDYWxpZm9ybmlh
MRYwFAYDVQQHEw1Nb3VudGFpbiBWaWV3MRwwGgYDVQQKExNNb3ppbGxhIENvcnBv
cmF0aW9uMRwwGgYDVQQLExNSZWxlYXNlIEVuZ2luZWVyaW5nMRwwGgYDVQQDExNN
b3ppbGxhIENvcnBvcmF0aW9uMTMwMQYJKoZIhvcNAQkBBCQwIhYgcmVsZWFzZStj
ZXJ0aWZpY2F0ZXNAbW96aWxsYS5jb20wggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAw
ggEKAoIBAQC0DJaUCP4FKWJWxh2t1mGtQfoMgVIq9czrbEMX3pH9QaEvHiY8HYY6
A9l4ph7Z3KE8rd9EW20eszvUvdyIcp0Wm3E4cbg7acsoOmcAbfgz0Me9Ua4NcEk1
sSTtYI5wf8PrMQh2m1lJkjFn4OW/FrrI+7NFDyX+//umYHYRIOq8ytGnv9OBKv96
6h3Mq13w8NVRID+SSgw3zfkiw8OnxIx0IlnN88aO1QvGpfSLTUqEUvCR4WJGrlpe
zWuozeJ1RukQvxt4NSfcYHhkvAzBxEZVENX+WI0EN7nLNv53odoiaVYauMNoUYFW
O71s6E9qpT1K2D5pGeVrJUDhbjy1cSoXAgMBAAGgMTAvBgkqhkiG9w0BCQ4xIjAg
MB4GA1UdEQQXMBWCE01vemlsbGEgQ29ycG9yYXRpb24wDQYJKoZIhvcNAQELBQAD
ggEBAIGPj4mMwAeSplfmSQ06OLME30UQjTJbVOEMdRbQP3NEA94wdQJJxY/EeLpU
yjsbEMXudBovJWcDwdl4ML5LtK8ePDZlLadjb23GGM/k9rT0xzPpfdaILGpw56pf
RMESpZP/wfQ5plGqgB3C+N00e8EX3EI0A2FO4gaCOuaK/OZLsuv9JvXm8MLEI3fh
A9WJItq1ofzbZHMpLq1DY4IbbjLT2OcKqPaO1xTTUmO2OE6xzz5kuAIcyXcdGvj4
rd4UOQt1ueBfuISGWm47Q9V1xoR4gOw4gkcct0OquUwC3X/DQCXBrb2n0eHnqLOU
6YlS70eKNBGcteA2xlxMXkj7PJM=
-----END CERTIFICATE REQUEST-----
```

File: tools/makecsr/make-ev-csr.go [hunk 1]
```diff
@@ -23,16 +25,20 @@ import (
...
-	// flag.StringVar(&cn, "cn", "Content Signing Intermediate", "CommonName of the Subject")
-	// flag.StringVar(&email, "email", "foxsec@mozilla.com", "Email of the Subject")
 	flag.Parse()
 
+	// We need the email address encoded as asn1.ia5
+	email_utf8 := "release+certificates@mozilla.com"
+	type ia5String struct {
+		A string `asn1:"ia5"`
+	}
+	email_ia5, err := asn1.Marshal(ia5String{email_utf8})
```
Comment (jvehent) [inline tools/makecsr/make-ev-csr.go]: I don't think you should marshal here. Instead, try setting the Value on line 76 to `ia5String{email_utf8}`

Comment (hwine) [reply tools/makecsr/make-ev-csr.go]: My understanding is that marshal does the actual encoding. Until then, it's a normal golang string with a tag.

I used [this tool](https://lapo.it/asn1js/#) to verify that the field is identified as ia5 encoding in the output "csr". (The output is not a valid csr, but it is valid asn1.)

I'll test my understanding later today.

Comment (jvehent) [reply tools/makecsr/make-ev-csr.go]: The call to `xz509.CreateCertificateRequest` later will take care of `asn1.Marshal` for the entire subject.

Comment (hwine) [reply tools/makecsr/make-ev-csr.go]: Nope. With the direct assignment, it decodes as:
```
SEQUENCE (2 elem)
          OBJECT IDENTIFIER 1.2.840.113549.1.9.1 emailAddress (PKCS #9. Deprecated, use an altName extension instead)
          UTF8String release+certificates@mozilla.com
```

With the call to marshal, it decodes as:
```
SEQUENCE (2 elem)
          OBJECT IDENTIFIER 1.2.840.113549.1.9.1 emailAddress (PKCS #9. Deprecated, use an altName extension instead)
          OCTET STRING (1 elem)
            SEQUENCE (1 elem)
              IA5String release+certificates@mozilla.com
```

But now I see I wasn't using the full output of the tool -- I'll go look for `altName` OID now

**Comment by jvehent** [comment]
@hwine what should we do with this PR?

**Comment by hwine** [comment]
diaf!!!!  I should make sure the docs are updated on the lessons learned -- 

@jvehent any better spot than Autograph? Seems like this was releng lore.

**Pull request closed without merging**